[staging] python: add pythonRemoveTestsDirHook

jonringer opened the pull request:
###### Motivation for this change
closes: #81482
```
Some packages do not have an explicit
find_packages(...,exclude=["tests"]) in thier
setup.py. This causes a tests module to be
installed, and can collide with other packages
which fail to filter it out as well.
```

This could potentially create problems if there ever was a "tests" package. But I don't think that will ever be the case with how prevalent a "tests" directory is.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

jonringer commented:
@GrahamcOfBorg build python38Packages.poetry

jonringer commented:
actually, i need ofborg to evaluate `python38.withPackages(ps: with ps; [poetry])`, not sure how to do that

adisbladis commented:
@jonringer I recently implemented the same fix for poetry2nix as a hook: https://github.com/nix-community/poetry2nix/blob/master/hooks/fixup-hook.sh.

It's better as a hook in regards to overrides.

FRidh commented:
I used to just build a sdist first, as I needed that anyway when distributing source code. IMO that's how we should do it when building from git.

jonringer commented:
this does solve the related issue:

```
[09:15:33] jon@jon-desktop /home/jon/projects/nixpkgs (avoid-test-dir)
$ nix-shell -p "with import ./. {}; python38.withPackages(ps: with ps; [poetry])"

[nix-shell:/home/jon/projects/nixpkgs]$
```

jonringer commented:
> I used to just build a sdist first, as I needed that anyway when distributing source code. IMO that's how we should do it when building from git.

I don't think we have a good mechanism to determine this, as both will hand us a `tar.gz`. I think going the hook route is fine if you want me to refactor to something similar to what @adisbladis posted

FRidh commented:
> I don't think we have a good mechanism to determine this, as both will hand us a tar.gz. 

Right, it would be a Nix function building a derivation, e.g. `buildSetuptoolsSdist`.

Commit:
python: add pythonRemoveTestsDirHook

This will automatically remove a top-level "tests"
directory from being installed.

jonringer commented:
I changed it to being a hook

Seeing as the "tests" package on pypi https://pypi.org/project/tests/ is defunct, I think it's safe that you will never want a top-level "tests" package

jonringer commented:
> Right, it would be a Nix function building a derivation, e.g. `buildSetuptoolsSdist`.

or use the pre-existing `format` switch, and just have `format = "sdist";` in an expression

FRidh commented on pkgs/development/interpreters/python/hooks/python-remove-tests-dir-hook.sh:
```diff
@@ -0,0 +1,15 @@
+    echo "Finished executing pythonRemoveTestsDir"
+}
+
+if [ -z "${dontRemoveTestsDir-}" ]; then
```
`dontUsePythonRemoveTestsDir` not very short but consistent

adisbladis replied:
I'll align `poetry2nix` with this naming too.

jonringer replied:
I think i renamed it, and forgot to update this. Thanks!

jonringer replied:
although, i forgot about the `Use` portion anyway :)

FRidh commented:
Now that we're talking about hooks, I started off a hook that would remove requirements from a wheel https://github.com/NixOS/nixpkgs/compare/master...FRidh:python-deps-hook, that way it should work regardless of where the dependency is listed. It's not functioning (need to look further into zipfile class) but it shows the idea.

jonringer commented:
> Now that we're talking about hooks, I started off a hook that would remove requirements from a wheel [master...FRidh:python-deps-hook](https://github.com/NixOS/nixpkgs/compare/master...FRidh:python-deps-hook), that way it should work regardless of where the dependency is listed. It's not functioning (need to look further into zipfile class) but it shows the idea.

For packages that over-pin, I think this is great, but it still leans more on having tests to verify that a package set is still valid.

jonringer commented:
with current changes, this has the expected behavior:
```
[11:00:19] jon@jon-desktop /home/jon/projects/nixpkgs (avoid-test-dir)
$ NIX_PATH=nixpkgs=$PWD nix-shell -p "python38.withPackages(ps: with ps; [poetry])"

[nix-shell:/home/jon/projects/nixpkgs]$ poetry --version
Poetry version 1.0.3

[nix-shell:/home/jon/projects/nixpkgs]$ cat pkgs/development/python-modules/poetry/default.nix | grep version
  version = "1.0.3";
```

jonringer commented:
@FRidh does this look good to you? I know that the python ecosystem is mostly your baby

FRidh commented:
I hope we could fix it at the source side instead of fixing up but I think we may have to do it this anyway.

Something I quickly gave a try. So far the issue have been the packages for which an sdist was built using poetry. Rebuilding an sdist for those requires poetry resulting in recursion. The main issue is with those packages built using poetry so I prefer to look at that deeper before using the hammer.
https://github.com/NixOS/nixpkgs/compare/master...FRidh:sdist?expand=1

FRidh commented:
Actually I do not want to spend time on this, let's merge.
